markovflow/models/variational_cvi_sde.py

```diff
@@ -0,0 +1,399 @@
+
+    The term E_{q}[||f_L - f_p||^2_{Σ^{-1}}] is termed as linearization loss as it can be interpreted as the metric for
+    the difference between the prior SDE and the linearized SDE along the current posterior.
+
+    The term E_{q}[(f_q - f_L){Σ^{-1}}(f_L - f_p)] is termed as the cross term.
+
+    """
+
+    def __init__(
+            self,
+            prior_sde: SDE,
+            time_grid: TensorType,
+            input_data: Tuple[TensorType, TensorType],
+            likelihood: Likelihood,
+            learning_rate=0.1,
+            prior_initial_state: Gaussian = None,
+            initial_posterior_path: Gaussian = None,
+    ) -> None:
+        """
+        :param prior_sde: Prior SDE over the latent states, x.
+        :param time_grid: Grid over time with shape ``batch_shape + [grid_size]``
+        :param input_data: A tuple containing the observed data:
+
+            * Time points of observations with shape ``batch_shape + [num_data]``
+            * Observations with shape ``batch_shape + [num_data, observation_dim]``
+
+        :param likelihood: A likelihood for the observations of the model.
+        :param learning_rate: The learning rate of the algorithm.
+        :param prior_initial_state: A Gaussian prior on the initial state.
+        :param initial_posterior_path: A Gaussian prior on the initial posterior path.
+
+        Note: Currently, only batch shape 1 is supported.
+        """
+        super().__init__()
+
+        self._likelihood = likelihood
+        self.time_grid = time_grid
+        self.prior_sde = prior_sde
+
+        self._observations_time_points = input_data[0]
+        self._observations = input_data[1]
+
+        self.output_dim = 1
+        self.state_dim = self._observations.shape[-1]
+
+        self.sites = UnivariateGaussianSitesNat(
+            nat1=Parameter(tf.zeros_like(self._observations_time_points)[..., None]),
+            nat2=Parameter(tf.ones_like(self._observations_time_points)[..., None, None] * -1e-10),
+            log_norm=Parameter(tf.zeros_like(self._observations_time_points)[..., None]),
+        )
+        self.sites_lr = learning_rate
+
+        self._initialize_initial_state_prior(prior_initial_state)
+        self._initialize_posterior_path(initial_posterior_path)
+
+        self.linearization_pnts = (tf.identity(self.fx_mus[:, :-1, :]), tf.identity(self.fx_covs[:, :-1, :, :]))
+        self._linearize_prior()
+
+        self.obs_sites_indices = tf.where(tf.equal(self.time_grid[..., None], self._observations_time_points))[:, 0][
+            ..., None]
+        self.dt = self.time_grid[1] - self.time_grid[0]
+
+    def _initialize_initial_state_prior(self, prior_initial_state: Gaussian):
+        """Initialize the prior on the initial state."""
+        if prior_initial_state is None:
+            prior_initial_state = Gaussian(mu=tf.zeros((1, self.state_dim), dtype=self._observations.dtype),
```

Comment:
Any reason not to store the gpflow Gaussian object? rather than the two moments?
Here you pass or build a gaussian to later extract and store the mean and covariance

Reply:
It's basically done this way because when these are used in `linearize_sde` method we will need to do casting and reshaping there so why not do it in the init. 

Now, when i think about it, linearize_sde can also take a Gaussian rather than `m` and `S`. :/